### DISCUSSION TITLE
Using Parallelism to Generate Discrete Gaussian Samples in OpenDP

### DISCUSSION THREAD

**Discussion by Lucas-Kong** [opendp/opendp#2135] [OpenDP Library] [answered]
Hello, I am trying to generate a sample of a discrete Gaussian distribution with a size of 1000 using the Python OpenDP library. I have imported opendp.prelude as dp and tried to use the dp.m.make_gaussian function. My questions are as follows:

1. I generated 1000 samples by evaluating the measurement using gaussian = dp.m.make_gaussian(*input_space, scale=sigma_sq_) and samples = [gaussian(0) for _ in range(size)] (where size = 1000). When I call the make_gaussian function, is there any way to generate samples using parallelism, which is supported by Rust, instead of using a for loop?

2. I am trying to generate a discrete Gaussian distribution that results in integers instead of a continuous Gaussian distribution that results in floating-point numbers. Is it sufficient to set the attribute of input_space as int? (i.e., input_space = (dp.atom_domain(T=int), dp.absolute_distance(T=int)))

Here is my code.
<code>
import opendp.prelude as dp
dp.enable_features('contrib')
input_space = dp.atom_domain(T=int), dp.absolute_distance(T=int)
gaussian = dp.m.make_gaussian(*input_space, scale=1.0)

size = 1000
from fractions import Fraction
import math
for ssqn,ssqd in [(5023 ,503), (1511, 151), (1500, 150), (751, 79), (10, 1), (1,1), (499, 503), (9041, 9041), (9000, 9000), (9049,9059), (9049, 9043), (1, 10), (307, 3001), (300, 3000), (53, 503)]:
    sigma_sq_ = math.sqrt(Fraction(ssqn, ssqd))
    gaussian = dp.m.make_gaussian(*input_space, scale=sigma_sq_)
    samples = [gaussian(0) for _ in range(size)]
    print(f"gauss_samples_{ssqn}_{ssqd}_{size}= {samples}")

Thank you.

**Comment by Shoeboxam** [comment, marked as answer]
Hi Lucas! 

1. From my profiling there's a fair amount of overhead in lots of repeated individuals calls to the mechanism that gets amortized when using the vector-valued gaussian mechanism.

```python
import opendp.prelude as dp
dp.enable_features('contrib')

input_space = dp.vector_domain(dp.atom_domain(T=int)), dp.l2_distance(T=int)
gaussian = dp.m.make_gaussian(*input_space, scale=1.0)

size = 1000
from fractions import Fraction
import math
for ssqn,ssqd in [(5023 ,503), (1511, 151), (1500, 150), (751, 79), (10, 1), (1,1), (499, 503), (9041, 9041), (9000, 9000), (9049,9059), (9049, 9043), (1, 10), (307, 3001), (300, 3000), (53, 503)]:
    sigma_sq_ = math.sqrt(Fraction(ssqn, ssqd))
    gaussian = dp.m.make_gaussian(*input_space, scale=sigma_sq_)
    samples = gaussian([0] * size)
    print(f"gauss_samples_{ssqn}{ssqd}{size}= {samples}")
```

Here I pushed the loop down into Rust.

While the Rust code doesn't use parallelism to generate these samples in a loop, I could make this adjustment.

2. make_gaussian internally always uses the discrete gaussian mechanism, even when you noise floats. It essentially uses the discrete gaussian mechanism as-is when T=int. If T=float, then the library does extra work to discretize first and convert back out to floats later.

Thanks for posting, let me know if you have any more questions or something still isn't clear.

**Reply by Shoeboxam** [reply]
I initially wrote this up but realized it wasn't relevant to your question. Leaving it here regardless, will work it into docs at some point.

When you switch the input space to `float`, then the library will first rationalize each float (where incrementing the numerator equivalently increments the float by one subnormal), adds noise to the numerator with the proportional scale, then converts back out. It looks nearly indistiguishable from continuous gaussian noise. 

```python
import opendp.prelude as dp
dp.enable_features('contrib')


input_space = dp.vector_domain(dp.atom_domain(T=float)), dp.l2_distance(T=int)
gaussian = dp.m.make_gaussian(*input_space, scale=1.0)

size = 1000
from fractions import Fraction
import math
for ssqn,ssqd in [(5023 ,503), (1511, 151), (1500, 150), (751, 79), (10, 1), (1,1), (499, 503), (9041, 9041), (9000, 9000), (9049,9059), (9049, 9043), (1, 10), (307, 3001), (300, 3000), (53, 503)]:
    sigma_sq_ = math.sqrt(Fraction(ssqn, ssqd))
    gaussian = dp.m.make_gaussian(*input_space, scale=sigma_sq_)
    samples = gaussian([0.] * size)
    print(f"gauss_samples_{ssqn}{ssqd}{size}= {samples}")
```

You might notice that such a fine discretization causes a slower runtime. If you want faster sampling, you can choose a less fine discretization:

```python
import opendp.prelude as dp
dp.enable_features('contrib')

size = 1000

input_space = dp.vector_domain(dp.atom_domain(T=float), size=size), dp.l2_distance(T=int)
gaussian = dp.m.make_gaussian(*input_space, scale=1.0)

from fractions import Fraction
import math
for ssqn,ssqd in [(5023 ,503), (1511, 151), (1500, 150), (751, 79), (10, 1), (1,1), (499, 503), (9041, 9041), (9000, 9000), (9049,9059), (9049, 9043), (1, 10), (307, 3001), (300, 3000), (53, 503)]:
    sigma_sq_ = math.sqrt(Fraction(ssqn, ssqd))
    gaussian = dp.m.make_gaussian(*input_space, scale=sigma_sq_, k=-50)
    samples = gaussian([0.] * size)
    print(f"gauss_samples_{ssqn}{ssqd}{size}= {samples}")
```

This time I set k=-50 (the gap between adjacent values in the support is now 2^-50). If k is left unset (as in the previous example), k defaults to -1074 (the gap between adjacent subnormal doubles is 2^-1074). When the discretization is more coarse (k is higher than -1074), then discretizing increases the sensitivity according to the length of the vector, thus requiring the domain to have known size. This info is passed into the input_space and used to correctly calibrate the privacy guarantee. All in all, a faster runtime/coarser support results in a slight increase in sensitivity and slightly less private output.
